### PR TITLE
solutions for week 5 ( 1639024365, 6260)

### PR DESCRIPTION
**Pull request by christianrietzsch**

### CODE REVIEW

**Comment by TomKnoden** [comment]
Please also provide the solutions to the given tasks, for example as part of your description.

File: Exercises/ChristianRietzsch/exercise_05.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+if(b:=lambda d,e:(q(g+g*b(f,e)for(g,f)in e[d])if e[d]else 0))
+and(i:=t(h))and(not v(h))and([w(a)({j[:-5]:[(k:=(y(l[:q([1for(m)in(l)if
+u(m)])])if(q([1for(m)in(l)if u(m)])!=0)else 1),x(l[o(p(k)):])())for(l)in
+s(x(z(z(n,ab),ac))("."))(", ")]if(ae)not in(n)else[]})for(j,n)in[s(x(o)())(ad)
+for(o)in s(i)("\n")]])else 0)
+
+assert(c({},open("data.txt","r"),lambda x:x.split,lambda y:y.strip,
+lambda x,y:x.replace(y,""),sum,lambda x:x.isdigit(),len,str,int,
+lambda x:x.read(),lambda x:x.close(),lambda x:x.update,"shiny gold",
+" bags"," bag"," contain ","no other bags") == 6260)
```
Comment (Tadd12) [inline Exercises/ChristianRietzsch/exercise_05.py:76]: Use proper indents to make the code readable

File: Exercises/ChristianRietzsch/exercise_05.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+for(string)in dstr.split("\n")]]))else 0)
+
+#compressed:
+def c(a,h,s,x,z,q,u,o,p,y,t,v,w,r,ab,ac,ad,ae):return(b(r,a)
+if(b:=lambda d,e:(q(g+g*b(f,e)for(g,f)in e[d])if e[d]else 0))
+and(i:=t(h))and(not v(h))and([w(a)({j[:-5]:[(k:=(y(l[:q([1for(m)in(l)if
+u(m)])])if(q([1for(m)in(l)if u(m)])!=0)else 1),x(l[o(p(k)):])())for(l)in
+s(x(z(z(n,ab),ac))("."))(", ")]if(ae)not in(n)else[]})for(j,n)in[s(x(o)())(ad)
+for(o)in s(i)("\n")]])else 0)
+
```
Comment (Tadd12) [inline Exercises/ChristianRietzsch/exercise_05.py:72]: Explain the variable names. What do they stand for?
Use proper indents to make the code readable

File: Exercises/ChristianRietzsch/exercise_05.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+else sum(val+val*count(ibag,cbag)
+for(val,ibag)in cbag[bag])))and(data:=open("data.txt","r"))
+and(dstr:=data.read())and(not data.close())
+and(_:=[pbag.update({bag[:-5]:[(ilen:=(int(cont[:sum([1for(cchar)
+in(cont)if cchar.isdigit()])])if(sum([1for(cchar)in(cont)
+if cchar.isdigit()])!=0)else 1),cont[len(str(ilen)):].strip())
+for(cont)in conts.replace(" bags","").replace(" bag", "").rstrip(".")
+.split(", ")] if"no other bags"not in conts else[]})
+for(bag,conts)in[string.strip().split(" contain ")
+for(string)in dstr.split("\n")]]))else 0)
```
Comment (Tadd12) [inline Exercises/ChristianRietzsch/exercise_05.py:63]: Use proper idents to make the code readable

File: Exercises/ChristianRietzsch/exercise_05.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        return 1
+
+def rec_count(item,complete_bag):
+    return (0 if not complete_bag[item] else
+    (sum(value+value*rec_count(bags,complete_bag) for value,bags in complete_bag[item])))
+
+def count_bags_inside(parsed_bag={}):
+    file = open("ppp-2023/data/input_bags.txt", "r")
+    data_str = file.read()
+    file.close()
```
Comment (Tadd12) [inline Exercises/ChristianRietzsch/exercise_05.py:40]: use a context manager to ensure the file being closed even if an error is thrown
```python
with open("ppp-2023/data/input_bag.txt", "r") as file:
    data_str = file.read()
``` 

> When a resource is local to a particular section of code, use a `with` statement to ensure it is cleaned up promptly and reliably after use. A try/finally statement is also acceptable.

File: Exercises/ChristianRietzsch/exercise_05.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    for string in data_str.split("\n"):
+        bag, contents = string.replace(" bags","").replace(" bag","").split(' contain ')
+        if("no other" not in contents):
+            parsed_bag[bag] = [(getNumber(bags),bags[len(str(getNumber(bags))) :]
+            .strip()) for bags in contents.rstrip(".").split(", ")]
```
Comment (Tadd12) [inline Exercises/ChristianRietzsch/exercise_05.py:45]: missing an indent

**Review by Tadd12 — COMMENTED** [review]
In summary:
- don't try to be cool by compressing python code. That's not something to be proud of!
- use context managers
- try to consider runtimer

No need to be sad, you will learn python conventions after working with it for a while

File: Exercises/ChristianRietzsch/exercise_05.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+def is_sum(check_num, preceding_nums):
+    for number_1 in preceding_nums:
+        for number_2 in preceding_nums:
+            if(number_1 != number_2 and (number_1 + number_2) == check_num):
+                return True;
+    return False;
+
+def find_number_exception(preceding_range, numbers) -> int :
+    for i in range(preceding_range, len(numbers)):
+        if(not is_sum(numbers[i], numbers[(i-preceding_range):i])):
```
Comment (joh-lin) [inline Exercises/ChristianRietzsch/exercise_05.py:10]: Moving the number validation to an outside function is a very good idea.

File: Exercises/ChristianRietzsch/exercise_05.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            return numbers[i];
+    return 0;
+
+file = open("ppp-2023/data/input_sequence.txt", "r")
+data_str = file.read()
```
Comment (joh-lin) [inline Exercises/ChristianRietzsch/exercise_05.py:15]: You could also have used file.readlines()

File: Exercises/ChristianRietzsch/exercise_05.py (hunk 1 above)
Comment (joh-lin) [inline Exercises/ChristianRietzsch/exercise_05.py:76]: beautiful 🤩

**Review by joh-lin — APPROVED** [review]

**Pull request merged**